src/Codegen/Codegen.hs

```diff
@@ -359,80 +374,83 @@ genYield :: GenFn [C.BlockItem]
->
->   // After the return, x is no longer initialized, so the following is
->   // undefined behavior:
->   x + 6
-
-To ensure this is cannot happen, we conservatively declare @x@ as a local
-variable in the activation record, so that its value is preserved between
-yields, even if this is not usually necessary. We leave it to later compiler
-passes to optimize this.
--}
+-- | Translate an SSM expression into a C expression and statements.
+--
+-- SSM IR is a side-effectful expression language, with two implications when
+-- translating to C:
+--
+-- 1.  every expression has a value (even if it is an uninhabited type), so this
+--    must be reflected in C; and
+-- 2.  some of the side effects in SSM cannot be implemented in C using expressions
+--    alone.
+--
+-- These two implications roughly translate to the @C.Exp@ and @[C.BlockItem]@ in
+-- @genExpr@'s return type. When we translate an SSM expression @e@:
+--
+-- > (val, stms) <- genExpr e
+--
+-- @val@ represents the C expression that corresponds to the value of @e@ upon
+-- evaluation, while @stms@ represents the list of preceding statements that
+-- compute @val@.
+--
+-- A further consideration upon point 2 is that SSM expressions may yield control
+-- at any point. Thus, the C expression returned by @genExpr@ must accommodate the
+-- step function suspending and resuming. For instance, consider the following SSM
+-- IR expression:
+--
+-- > (let x = 3 in x) + (wait r; 6)
+--
+-- The @x@ in the let-binding in the left operand cannot just be a local variable
+-- in the step function, because it would be "uninitialized" by the yield in the
+-- right operand:
+--
+-- >   // let x = 3 in x
+-- >   // stms:
+-- >   int x = 3;
+-- >   // exp: x
+-- >
+-- >   // (wait r; 6)
+-- >   // stms:
+-- >   ssm_sensitize(r);
+-- >   actg->pc = N;
+-- >   return;
+-- > case N:
+-- >   ssm_desensitize(r);
+-- >   // exp: 6
+-- >
+-- >   // After the return, x is no longer initialized, so the following is
+-- >   // undefined behavior:
+-- >   x + 6
+--
+-- To ensure this is cannot happen, we conservatively declare @x@ as a local
+-- variable in the activation record, so that its value is preserved between
+-- yields, even if this is not usually necessary. We leave it to later compiler
+-- passes to optimize this.
 genExpr :: I.Expr I.Type -> GenFn (C.Exp, [C.BlockItem])
 genExpr (I.Var n (I.TBuiltin (I.Arrow _ _))) =
   -- NOTE: Any identifiers of I.Arrow type must refer to a (global) function, so
   -- we just return a handle to its enter function.
   return ([cexp|$id:(enter_ n)|], [])
 genExpr (I.Var n _) = do
-  -- isLocal <- isLocalVar n -- TODO: check for global vs local variable
-  return ([cexp|$id:acts->$id:n|], [])
-genExpr (I.Data _ _             ) = nope
-genExpr (I.Lit  l t             ) = genLiteral l t
+  Just v <- M.lookup n <$> gets fnVars
+  return (v, [])
+genExpr (I.Data tg _) = do
+  info <- gets adtInfo
+  let Just initExpr = M.lookup tg (intInit info) in return (initExpr, [])
```

Comment:
I think this could be more clearly written as:

```haskell
do
  Just initExpr <- M.lookup tg . intInit <$> gets adtInfo
  return (initExpr, [])
```

Reply:
I reduced everything into a single line per your suggestion, but this change produced a table lookup error at runtime. I think the problem has something to do with trying to map the lookup call over a monadic table, or maybe the mix of monads Maybe/GenFn. I was tried to clean up the code a little while still preserving functionality:
```
info <- intInit <$> gets fnAdtInfo
let Just initExpr = M.lookup tg info
return (initExpr, [])
```